SOLR-15955: Update Jetty dependency to 10

markrmiller opened the pull request:
https://issues.apache.org/jira/browse/SOLR-15955

Summary:
* Upgrades to Jetty 10.0.12
* dropwizard metrics 4.2.12 for `dropwizard-metrics9` -> `dropwizard-metrics10`
* log4j 2.19.0 and slf4j 2.0.3
* for s3mock specifically upgrade spring-boot 2.5.14 and spring 5.3.23 to handle Jetty 10

madrob commented on .github/workflows/manual.yml:
```diff
@@ -0,0 +1,21 @@
+name: Scheduled Merge Remote Action
```
I don't think we want this? Might be surprising for some folks...

markrmiller replied:
unintended

markrmiller replied:
strange, don't even see it locally

madrob commented on solr/modules/s3-repository/src/test/org/apache/solr/s3/S3BackupRepositoryTest.java:
```diff
@@ -246,7 +266,7 @@ private void doRandomAccessTest(String content, int position) throws Exception {
     try (S3BackupRepository repo = getRepository()) {
       // Open an index input on a file
-      pushObject("/my-repo/content", content);
+      pushObject("my-repo/content", content);
```
@HoustonPutman I recall you fighting with S3 and Mock-S3 and how it handled directories...

markrmiller replied:
something changed with the version change of the mock lib

madrob commented on solr/modules/s3-repository/src/test/org/apache/solr/s3/S3ReadWriteTest.java:
```diff
@@ -91,7 +91,7 @@ public void testDirectoryLength() throws Exception {
             S3Exception.class,
             () -> client.length("/directory"));
-    assertEquals("Path is Directory", exception.getMessage());
+    assertTrue(exception.getMessage(), exception.getMessage().contains("Path is Directory"));
```
nit: `assertThat` would get us a nice error message for free

madrob commented on solr/core/src/java/org/apache/solr/request/json/RequestUtil.java:
```diff
@@ -68,8 +69,12 @@ public static void processParams(SolrRequestHandler handler, SolrQueryRequest re
         String contentType = cs.getContentType();
-        if (contentType==null || !contentType.contains("/json")) {
+        if (contentType==null || (!contentType.contains("/json") && !contentType.contains(
+            BinaryResponseParser.BINARY_CONTENT_TYPE))) {
```
I saw that you switched a bunch of places away from BINARY_CONTENT_TYPE to multi-part -- does Jetty 10 not like octet-stream? What's the motivation?

markrmiller replied:
3 or 4 tests fail due to the search handler and it's content header check.

madrob commented on solr/modules/s3-repository/src/test/org/apache/solr/s3/S3BackupRepositoryTest.java:
```diff
@@ -320,17 +340,113 @@ protected URI getBaseUri() throws URISyntaxException {
+              }
+            }
+          },
+          new java.security.SecureRandom());
```
Should we use NotSecurePseudoRandom here?

markrmiller replied:
Maybe, but unrelated to the jetty changes.

madrob commented on solr/modules/s3-repository/src/test/org/apache/solr/s3/S3BackupRepositoryTest.java:
```diff
@@ -320,17 +340,113 @@ protected URI getBaseUri() throws URISyntaxException {
     }
   }
+
+  public S3Client createS3ClientV2() {
```
Is this copied directly from adobe's S3MockStarter? If so, I'd like to see it in a separate file that we can properly attribute (and keep up to date periodically as needed)

markrmiller replied:
This should not be a change in this issue.

markrmiller replied:
Actually, I'm going to have check if that is still necessary or not. Was working on making some different version combo work properly.

madrob replied:
I don't think we need this - tried running tests locally without it and using S3_MOCK_RULE to build the client directly and things seemed fine.

risdenk replied:
this was addressed in later update.

madrob commented on solr/solrj/src/java/org/apache/solr/client/solrj/util/SolrBasicAuthentication.java:
```diff
@@ -35,7 +35,7 @@
 
   public SolrBasicAuthentication(String user, String password) {
-    this.value = "Basic " + B64Code.encode(user + ":" + password, StandardCharsets.ISO_8859_1);
+    this.value = "Basic " + Base64.getEncoder().encodeToString((user + ":" + password + "").getBytes(StandardCharsets.ISO_8859_1));
```
nit: extra `+ ""` in here

madrob commented on solr/test-framework/src/java/org/apache/solr/cloud/api/collections/AbstractBackupRepositoryTest.java:
```diff
@@ -103,6 +104,11 @@ public void testCanDistinguishBetweenFilesAndDirectories() throws Exception {
             repo.createDirectory(nonEmptyDirUri);
             addFile(repo, nestedFileUri);
 
+            String[] list = repo.listAll(repo.resolve(nonEmptyDirUri));
```
left over from debugging?

markrmiller replied:
yup

markrmiller commented:
Thanks for taking a look. I have a n update I'll push I'll push tomorrow.

markrmiller commented on solr/modules/s3-repository/build.gradle:
```diff
@@ -39,6 +39,10 @@ dependencies {
     runtimeOnly (group: 'com.fasterxml.woodstox', name: 'woodstox-core')
     runtimeOnly (group: 'org.codehaus.woodstox', name: 'stax2-api')
 
+    implementation 'org.springframework.boot:spring-boot'
```
This was already pulled in, I had to control certain versions. A few changes since, I'll see what's still required.

madrob replied:
That is to say, what's wrong with using the existing s3mock dependency resolution

markrmiller replied:
That creates the Jetty 10 issues.

markrmiller commented:
A couple of those are extra lines that came in when I merged to a clean repo (test policy and s3 build.gradle I believe).

I'll get the follow up pushed tomorrow - I rebased yesterday and after resolving new conflicts, tests were flakier (a handful failed first run) and one test failed consistently (managed storage test I believe). So I held back to look into that a bit, but in the in end, it seems I see the same behavior on a clean checkout, so should not be related this work.

markrmiller commented on solr/modules/hdfs/src/test/org/apache/solr/cloud/hdfs/HdfsTestUtil.java:
```diff
@@ -208,6 +216,32 @@ public static MiniDFSCluster setupClass(String dir, boolean safeModeTesting, boo
     MiniDFSCluster dfsCluster = dfsClusterBuilder.build();
     HdfsUtil.TEST_CONF = getClientConfiguration(dfsCluster);
+    
+    // nocommit TODO remove
```
unrelated tweaks for hdfs tests since I was running them a fair bit when trying for jetty 11

markrmiller commented on gradle/testing/randomization/policies/solr-tests.policy:
```diff
@@ -38,6 +38,13 @@ grant {
+  // hdfs name node uses jetty, ends up with a JarResource in AllowedResourceAliasChecker#doStart
+  // which ends up needing to be able to read '/', see HttpServer2 - there is a comment I saw that
+  // indicated there may be a way to supply a custom AliasChecker
+  permission java.io.FilePermission "/", "read";
```
10.0.8 breaks the S3 Spring dependency again. I'll see if a different version works out.

markrmiller commented:
That was a bit more effort than expected, but updated jetty from 10.0.7 to 10.0.8. That removes the need for the security policy read permission for "/".

Rebased to latest main, let's see if these build checks pass, did a couple test runs without issue.

madrob commented on solr/modules/hadoop-auth/src/test/org/apache/solr/client/solrj/impl/Krb5HttpClientUtils.java:
```diff
@@ -36,10 +36,11 @@
     http2Client
         .getProtocolHandlers()
         .put(new WWWAuthenticationProtocolHandler(http2Client.getHttpClient()));
+    http2Client.setAuthenticationStore(authenticationStore);
```
We call this twice?

markrmiller replied:
came in during rebase reconciliation I think, back to one

madrob commented on solr/modules/s3-repository/build.gradle:
```diff
@@ -41,21 +41,25 @@ dependencies {
     exclude group: 'io.netty', module: '*'
-    exclude group: 'commons-io', module: 'commons-io'
+    exclude group: 'software.amazon.awssdk', module: 'netty-nio-client'
     exclude group: 'commons-codec', module: 'commons-codec'
+    exclude group: 'commons-io', module: 'commons-io'
     exclude group: 'org.apache.commons', module: 'commons-lang3'
-    exclude group: 'software.amazon.awssdk', module: 'netty-nio-client'
```
The rearranging here looks weird

markrmiller replied:
Not really sure why that is there, probably conflict merging over time and changes.

madrob commented on versions.lock:
```diff
@@ -259,9 +258,9 @@ org.ow2.asm:asm:9.2 (4 constraints: 4931393a)
-org.slf4j:jul-to-slf4j:1.7.36 (2 constraints: 1723911a)
-org.slf4j:slf4j-api:1.7.36 (38 constraints: 9dd389ac)
+org.slf4j:jcl-over-slf4j:1.7.35 (3 constraints: 05188eb8)
+org.slf4j:jul-to-slf4j:1.7.35 (2 constraints: 1723911a)
+org.slf4j:slf4j-api:1.7.35 (57 constraints: a71af872)
```
Downgrade?

madrob commented on gradle/solr/force-versions.gradle:
```diff
@@ -0,0 +1,35 @@
+project(":solr").subprojects {
+  plugins.withType(JavaPlugin) {
+    dependencies {
+      implementation enforcedPlatform('org.slf4j:slf4j-api:1.7.35')
```
Should we set this version in version.props like the rest of our libraries?

markrmiller replied:
I think this was done this way because you cannot easily do that here, or I assume Dawid would have, but I'll see what happens.

markrmiller replied:
Think it actually works.

madrob commented on solr/core/src/java/org/apache/solr/request/json/RequestUtil.java:
```diff
@@ -68,8 +69,12 @@ public static void processParams(SolrRequestHandler handler, SolrQueryRequest re
       String[] jsonFromParams = map.remove(JSON);  // params from the query string should come after (and hence override) JSON content streams
 
       for (ContentStream cs : req.getContentStreams()) {
+        // if BinaryResponseParser.BINARY_CONTENT_TYPE, let the following fail below - we may have adjusted the content without updating the content type
+        // problem in this case happens in a few tests, one seems to happen with kerberos and remote node query (HttpSolrCall's request proxy)
```
It took a bit of digging, but I finally figured it out. When Jetty receives a request with no body content, it creates a placeholder zero-length application/octet-stream content object. Then when it gets a redirect (like an explicit redirect handler, or some authentication plugins cause), it copies that body to the new request. When the new request gets to our search handlers, we see the body and save it as a stream. So I think that we could check for that in `SolrRequestParsers.parseParamsAndFillStreams` - the combination of a GET request with a zero length octet-stream body seems like a safe instance to drop the stream. Maybe we could drop all the stream bodies for all get requests, I'm not sure. This could be a bug in Jetty, but I haven't figured out a simple repro case for them yet.

markrmiller replied:
Good find, and that does jive with what was happening when I found the fail I saw here pop up on the mailing list in the past.

So I think the diagnosis makes sense, and the change sounds reasonable high level, but I am concerned that it requires that you can inspect the stream or count on the content-size header to properly describe the stream. You cannot generally do that though. The size is not required, usually not given in Solr, and you can't reasonable inspect a stream in a webapp/servlet unless you buffer it, which is also problematic. Given we are talking about a steram on a GET request, maybe somehow you can get away with this, but it seems pretty fragile?

markrmiller replied:
Hmmm, it shouldn't save the steam content I hope, that would prevent actual streaming. But in either case, that does remind me: I think there is a tiny little buffer that is used for content type auto detect in RequestParsers...we would know if a stream is empty or not there.

markrmiller commented on solr/core/build.gradle:
```diff
@@ -90,6 +90,7 @@ dependencies {
   implementation 'org.apache.httpcomponents:httpcore'
 
   compileOnlyApi 'javax.servlet:javax.servlet-api'
+  compileOnlyApi 'org.eclipse.jetty.toolchain:jetty-servlet-api'
```
> I don't think we need this 

I tried without it and it failed with a client/server using SSL mismatch before putting this iteration up. I can take one more look at it.

> I think having both of these is incorrect? 

I never actually added that myself, it did exist when I started, it was removed in a rebase I did, and then, and if I remove it now, analyzeClassesDependencies fails for me with (I'm digging around a bit, not sure why it existed, or why it could be removed but then not in this issue):


FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':solr:modules:jwt-auth:analyzeClassesDependencies'.
> Dependency analysis found issues.
  usedUndeclaredArtifacts
   - org.eclipse.jetty.toolchain:jetty-servlet-api:4.0.6@jar


Execution failed for task ':solr:modules:jwt-auth:analyzeClassesDependencies'.
> Dependency analysis found issues.
  usedUndeclaredArtifacts
   - org.eclipse.jetty.toolchain:jetty-servlet-api:4.0.6@jar

markrmiller replied:
Not sure what you want to do here. If I remove that, analyzeClassDependencies will fail.

Here is the details on that toolchain module, appears to be part of what is needed to build Jetty: https://github.com/eclipse/jetty.toolchain

Here is a dependency insight call on that lib:

org.eclipse.jetty.toolchain:jetty-servlet-api:4.0.6
+--- project :
|    \--- runtimeClasspath
+--- org.eclipse.jetty:jetty-rewrite:10.0.8
|    +--- runtimeClasspath
|    \--- project : (requested org.eclipse.jetty:jetty-rewrite:{strictly 10.0.8}) (*)
\--- org.eclipse.jetty:jetty-server:10.0.8
     +--- runtimeClasspath
     +--- project : (requested org.eclipse.jetty:jetty-server:{strictly 10.0.8}) (*)
     +--- io.dropwizard.metrics:metrics-jetty9:4.1.5 (requested org.eclipse.jetty:jetty-server:9.4.27.v20200227)
     |    +--- runtimeClasspath
     |    \--- project : (requested io.dropwizard.metrics:metrics-jetty9:{strictly 4.1.5}) (*)
     +--- org.eclipse.jetty:jetty-rewrite:10.0.8 (*)
     +--- org.eclipse.jetty:jetty-security:10.0.8
     |    +--- runtimeClasspath
     |    +--- project : (requested org.eclipse.jetty:jetty-security:{strictly 10.0.8}) (*)
     |    \--- org.eclipse.jetty:jetty-servlet:10.0.8
     |         +--- runtimeClasspath
     |         +--- project : (requested org.eclipse.jetty:jetty-servlet:{strictly 10.0.8}) (*)
     |         \--- org.eclipse.jetty:jetty-webapp:10.0.8
     |              +--- runtimeClasspath
     |              +--- project : (requested org.eclipse.jetty:jetty-webapp:{strictly 10.0.8}) (*)
     |              \--- org.eclipse.jetty:jetty-deploy:10.0.8
     |                   +--- runtimeClasspath
     |                   \--- project : (requested org.eclipse.jetty:jetty-deploy:{strictly 10.0.8}) (*)
     +--- org.eclipse.jetty:jetty-alpn-server:10.0.8
     |    +--- runtimeClasspath
     |    +--- project : (requested org.eclipse.jetty:jetty-alpn-server:{strictly 10.0.8}) (*)
     |    \--- org.eclipse.jetty:jetty-alpn-java-server:10.0.8
     |         +--- runtimeClasspath
     |         \--- project : (requested org.eclipse.jetty:jetty-alpn-java-server:{strictly 10.0.8}) (*)
     \--- org.eclipse.jetty.http2:http2-server:10.0.8
          +--- runtimeClasspath
          \--- project : (requested org.eclipse.j

risdenk replied:
This is addressed now

markrmiller commented:
> I don't think we need this - tried running tests locally without it and using S3_MOCK_RULE to build the client directly and things seemed fine.

I actually took things out one by one from this class on the last update, and I left this because the tests passed as I removed up to this point and then failed when I removed this. But I tried again, and it seems to have passed, so perhaps some stale result issue I saw. Looks like it can come out if we go by your result and my latest.

risdenk commented:
Is there more to do here? I know @anshumg asked that Jetty 9.4.x upgrade be superseded by this change. There have been a few more Jetty 9.4.x and 10.x releases.

madrob commented:
@elyograg FYI

markrmiller commented:
I can put up a more recent rebase.

The main reason this has sat around is a small but prominent back compat issue. 

We ship existing jetty.xml config with the use of a id tag that has always been incorrect and should be refid. In Jetty 10 that will hard fail. We also use methods on the gzip handler that were previously deprecated and then removed in 10. That will hard fail. 

As a result, with a default, clean Solr 9 release, without some potential solution, if you updated to Jetty 10 on a 9.x point release, Solr would fail to start on an existing install without a couple small, simple tweaks to the jetty xml config. That kind of break would be very abnormal for a point release.

The only potential workaround I’ve thought up so far is a kind of ugly, have the start script always check for this and and attempt to manually correct it, maybe only if the user hasn’t modified the files or something. But that’s far from a great solution.

epugh commented:
We've only shipped Sole 9.0...   Which I suspect means that very few people have actually upgraded from the 8.x line...     This back compat issue will become more of a problem from 9.1 and later as more people adopt Solr 9.   

Is modifying Jetty.xml a common thing to do?  And if it is, then maybe the release notes will warn them?   Be a shame to wait another 18 months till Solr 10 comes out just to finally get this done....

markrmiller commented:
I don’t know that modifying jetty.xml is that common. If the only case you were affected is if you had, I would just say release note it.

The problem is, you don’t need to have modified it. The version we shipped with won’t work, so you would have to overwrite that file on upgrade to avoid the issue.


If that is the standard upgrade process being used, then perhaps this is less of an issue.

What I imagine is a typical install layout that includes your Solr core directories and indexes and a point release upgrade would be commonly done by leaving your config and directories in place and just replacing the binary jars.


I don’t point release upgrade though, so maybe it’s normal to do a process that blows the jetty config files away?

Of course you could release now that you need to.  My concern is that to my knowledge we have not had a release note that if not followed means a standard upgrade process would result in Jetty failing to start.

epugh commented:
Yeah, I think it depends on how people do upgrades.  I *never* do an inplace upgrade without having tested it in Dev first, and I don't do inplace if I can help it!  

Is there anyway around this upgrade issue?  If not, is this worth a dev mailing list post to get some consensus on?

janhoy commented:
I'd say jetty.xml is an implementation detail and officially we don't need to give back-compat guarantees for it, and we should be allowed to assume that users do not hand-edit those files. In upgrade notes we can highlight that upgrading to 9.1 will replace jetty.xml, and users who rely on customizations to it will need to re-apply customizations after upgrade.

+1 to get Jetty 10 out in 9.x.

janhoy commented:
> I don’t point release upgrade though, so maybe it’s normal to do a process that blows the jetty config files away?

Docker users will always have a complete new image of course. Our official upgrade guide recommends taking down each node and doing a fresh install of the new version (https://solr.apache.org/guide/solr/latest/deployment-guide/upgrading-a-solr-cluster.html#upgrade-process). 

The Linux install script unpacks each version separately and uses a symlink to `/opt/solr` to isolate (https://solr.apache.org/guide/solr/latest/deployment-guide/taking-solr-to-production.html#service-installation-script).

So imo the project recommends a "fresh" install as best practice. Only in exceptional cases such as patching log4j jars do we recommend people to do in-place updates of individual jar files in an install. We can stress this further in the upgrade notes for 9.1.

epugh commented:
@markrmiller how can I help to get this ready to merge?   Or am I jumping the gun based on @janhoy +1 comment ?

If you don't want to write the release notes, I could take a stab at it.

elyograg commented:
I have done some manual edits on occasion to jetty.xml ... either to increase the header size limit so a larger query string can be handled with GET, or to enable request logging.  I expect that there will be a new jetty.xml when I upgrade -- I use the service installer, which means that the upgrade will create an entirely new program directory and update the symlink.  If instead of upgrading, I reinstall the same version with the -f flag, I think that would probably overwrite jetty.xml ... but I have never tried installing the same version.

I do not know what typically happens when upgrading a docker setup, as I have never installed Solr that way.  Does it create a new container that utilizes the same solr home that the old container was connected to?  I suspect that jetty.xml edits are never propagated to the new version when doing an upgrade in a docker setup, the same as upgrading a non-docker setup with the service installer script.

What would it realistically take to upgrade to Jetty 11 instead of 10?  Is that even possible without an extensive overhaul?  I see on the jetty website that 11 uses the new jakartaee.servlet.* packages and servlet API 5, while Jetty 10 uses the same javax.servlet.* packages as 9, and targets servlet API 4.  I would hope that version 11 has shim classes that allow webapps made for 9 to run.

elyograg commented:
We might want to provide mechanisms to make the most common jetty.xml edits configurable, which mostly eliminates that problem.  It is my opinion that request logging should be turned on by default if there is a way to have them automatically rotate and prune, but I am not going to insist on that color for the bikeshed.

janhoy commented:
> It is my opinion that request logging should be turned on by default

You should read up on the 9.0 release :) https://solr.apache.org/guide/solr/latest/deployment-guide/configuring-logging.html#request-logging ; https://issues.apache.org/jira/browse/SOLR-14142

epugh commented:
Agreed on making some of the configs in Jetty more accessible without changing files ;-).  Let's get to Jetty 10.   I'd love to see a list of what you might want to make more configurable...  CORS via Jetty filters is one that comes to mind for me...

markrmiller commented:
I’m happy to treat jetty.xml as an implantation detail, really only concerned with the semi odd shock of not being able to start up with an existing, unmodified file. If people are okay with a release note on that, I have no problem with it though.

I’ll rebase the patch again.

epugh commented:
A CVE, recorded as SOLR-16332, just showed up that would be fixed by this upgrade ;-).

epugh commented:
Anything I can do to help @markrmiller ?   I'm happy to take a stab at writing the Release Note for this...

risdenk commented:
> I’m happy to treat jetty.xml as an implantation detail, really only concerned with the semi odd shock of not being able to start up with an existing, unmodified file. If people are okay with a release note on that, I have no problem with it though.

I'm of the opinion that jetty.xml is implementation detail and release note is good enough. I can help merge main into this again see where we are with tests.

risdenk commented on solr/solrj/build.gradle:
```diff
@@ -75,6 +76,7 @@ dependencies {
   testImplementation 'commons-io:commons-io'
 
   testImplementation 'javax.servlet:javax.servlet-api'
```
Not sure we need both `'javax.servlet:javax.servlet-api'` and `'org.eclipse.jetty.toolchain:jetty-servlet-api'`? I need to check this.

risdenk replied:
I see there is a comment thread elsewhere on this in the PR.

risdenk replied:
This is addressed now.

risdenk commented on versions.lock:
```diff
@@ -230,30 +230,30 @@ org.codehaus.janino:commons-compiler:3.1.8 (2 constraints: 2f1983ec)
+org.eclipse.jetty:jetty-alpn-java-server:10.0.8 (1 constraints: 3b052f3b)
+org.eclipse.jetty:jetty-alpn-server:10.0.8 (2 constraints: 9916c323)
+org.eclipse.jetty:jetty-client:10.0.8 (2 constraints: c31a2343)
+org.eclipse.jetty:jetty-continuation:9.4.42.v20210604 (1 constraints: 721b9cb8)
```
This needs to be fixed.

risdenk replied:
This is addressed now.

risdenk reviewed (changes requested):
Working through some of the main merge issues locally related to some of the new jersey dependencies.

risdenk commented:
Well this was unexpected first time after fixing the static build stuff:

```
./gradlew clean
...
./gradlew check -Pvalidation.errorprone=true
BUILD SUCCESSFUL in 20m 50s
586 actionable tasks: 198 executed, 388 up-to-date
```

risdenk commented on solr/test-framework/build.gradle:
```diff
@@ -57,12 +57,12 @@ dependencies {
   implementation 'com.google.guava:guava'
   implementation 'io.dropwizard.metrics:metrics-core'
   implementation 'io.dropwizard.metrics:metrics-jetty9'
+  implementation 'org.eclipse.jetty.toolchain:jetty-servlet-api'
```
Move this back to line 65...

risdenk replied:
addressed in f5e8f3b2956ed5c4f024f8a10f52ee9a3b912e11

risdenk commented on versions.props:
```diff
@@ -68,5 +68,7 @@ org.openjdk.jmh:*=1.32
 org.quicktheories:quicktheories=0.26
 org.semver4j:semver4j=2.1.1
 org.slf4j:*=1.7.36
+org.springframework.boot:spring-boot=2.5.10
```
this needs a * to make sure all the spring boot gets updated.

risdenk replied:
addressed in 93937d7ca8a20409a05229466c43ffb98ecefd72

risdenk commented on solr/modules/s3-repository/build.gradle:
```diff
@@ -77,8 +79,9 @@ dependencies {
   testImplementation 'commons-io:commons-io'
   testImplementation 'org.apache.commons:commons-lang3'
-
+  
```
Fix this...

risdenk replied:
addressed in 133f4a86f45e871bcc4aae6af7f6ae17e22ca2b7

risdenk reviewed (changes requested):
Few minor things found upon review

risdenk commented on versions.lock:
```diff
@@ -369,8 +367,8 @@ org.apache.kerby:kerb-common:1.0.1 (2 constraints: a51841ca)
 org.apache.kerby:kerb-identity:1.0.1 (1 constraints: 5f0cb602)
 org.apache.kerby:kerb-server:1.0.1 (1 constraints: d10b65f2)
 org.apache.kerby:kerb-simplekdc:1.0.1 (1 constraints: dc0d7e3e)
+org.apache.tomcat.embed:tomcat-embed-el:9.0.63 (1 constraints: d01553cf)
```
due to https://github.com/spring-projects/spring-boot/issues/24744

risdenk replied:
replaces jakarta-el

dsmiley replied:
Where are we using this?

risdenk replied:
its the same question about jakarata-el that is replaced. s3mock uses spring boot and only in tests.

risdenk commented:
After most recent changes :)

```
./gradlew check -Pvalidation.errorprone=true
BUILD SUCCESSFUL in 20m 58s
586 actionable tasks: 196 executed, 390 up-to-date
```

risdenk commented:
Shawn Heisey correctly pointed out to me that metrics-jetty9 was still being used. Upgraded to metrics-jetty10 in 
3281463.

That made me go back and check other Jetty 9 references and found dtd should be updated. Fixed in d6a4c0d6f72ae7195b1233f6421279eb6a5c2baf

risdenk commented on solr/server/etc/jetty.xml:
```diff
@@ -178,7 +178,7 @@
                </Item>
                <Item>
-                 <New id="InstrumentedHandler" class="com.codahale.metrics.jetty9.InstrumentedHandler">
+                 <New id="InstrumentedHandler" class="com.codahale.metrics.jetty10.InstrumentedHandler">
```
Ugh @epugh  found that this is broken - it needs to be `io.dropwizard.metrics.jetty10.InstrumentedHandler`

risdenk replied:
Fixed in 1c9ea3ad9ac98cc133163f3e36c643ca567009a6

risdenk commented on versions.props:
```diff
@@ -17,14 +17,13 @@ commons-cli:commons-cli=1.4
 commons-collections:commons-collections=3.2.2
 commons-io:commons-io=2.11.0
-io.dropwizard.metrics:*=4.1.5
+io.dropwizard.metrics:*=4.2.12
```
Needed for metrics-jetty10

risdenk commented on versions.props:
```diff
@@ -68,5 +68,7 @@ org.openjdk.jmh:*=1.32
 org.quicktheories:quicktheories=0.26
 org.semver4j:semver4j=2.1.1
 org.slf4j:*=1.7.36
+org.springframework.boot:spring-boot*=2.5.14
+org.springframework:spring*=5.3.23
```
Needed for s3mock?

risdenk reviewed (approved):
```
./gradlew clean
...
BUILD SUCCESSFUL in 8s
33 actionable tasks: 31 executed, 2 up-to-date

./gradlew check -Pvalidation.errorprone=true
...
BUILD SUCCESSFUL in 21m 18s
586 actionable tasks: 534 executed, 52 up-to-date

./gradlew integrationTests
...
BUILD SUCCESSFUL in 6m 59s
77 actionable tasks: 60 executed, 17 up-to-date
```

dsmiley commented on gradle/solr/force-versions.gradle:
```diff
@@ -0,0 +1,35 @@
+ */
+
+
+// Force slf4j version so that it does not get auto promoted from 1x to 2-alpha due to Jetty dependency.
```
The technique here is new to me.  I see enforcedPlatform documented [here](https://docs.gradle.org/current/userguide/platforms.html#sub:bom_import).

Okay I guess.  It seems like this is a little-known solution to the problem we are solving.

risdenk replied:
So I know it was added a LONG time ago by dweiss during the initial gradle migration:
* https://github.com/apache/solr/commit/ca8661bc3ae4cf5076f5df6412f04f78350e238c
* https://github.com/apache/solr/commit/73e8b49f0d077e4c85d3d5c726d1acc0c6d2ac69

It was removed in SOLR-15987 
 since we were able to get to a consistent slf4j version again: https://github.com/apache/solr/commit/d9373b418474d28676e3812cf42e33d7b7f51c51

risdenk replied:
Addressed in 14b141e3a9f2842f7173d14ba05a8a625cb46bd2 - this fixes the slf4j logging stuff too you ran into. more details in separate comment in this PR.

risdenk replied:
Details in https://github.com/apache/solr/pull/585#issuecomment-1288324958 about slf4j-api issue.

dsmiley reviewed (changes requested):
I tried to do a quick manual run of this.
I did "gw assemble", CD'ed over there, then ran it in the foreground.
There is a serious problem with the logging configuration / JARs.  Slf4j complains and then there is no further logging output.  Solr is actually running though.
```
SLF4J: No SLF4J providers were found.
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#noProviders for further details.
SLF4J: Class path contains SLF4J bindings targeting slf4j-api versions prior to 1.8.
SLF4J: Ignoring binding found at [jar:file:/Users/dsmiley/SFDCDev/solr_main/solr/packaging/build/solr-10.0.0-SNAPSHOT/server/lib/ext/log4j-slf4j-impl-2.17.2.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#ignoredBindings for an explanation.
```

risdenk commented:
> SLF4J: Class path contains SLF4J bindings targeting slf4j-api versions prior to 1.8.

Hmmm - thats concerning..... we have slf4j 1.7.x in versions.lock and no other slf4j according to versions.lock. So somehow slf4j > 1.7 is getting included.

I'll take a closer look and see what is going on.

risdenk commented:
I found that `solr/packaging/build/solr-10.0.0-SNAPSHOT/server/lib/slf4j-api-2.0.0.jar` exists - which is super weird. going to keep digging

risdenk commented:
Ok so slf4j-api is used by jetty classes now. serverLib and jettyClientImplementation configurations were being used to determine which dependencies end up in which directories. Since Jetty was pulling in slf4j-api 2.0 it was getting put in a few places it shouldn't. 

I fixed this by:
* Upgrading slf4j-api to 2.0 in Solr so there aren't multiple different versions floating around
* excluding slf4j-api from jettyClientImplementation dependencies since then slfj4j-api doesn't end up in WEB-INF

This ensures a single slf4j-api version and logging works now :)

risdenk commented:
Need to address this:

```
2022-10-24 02:17:24.106 WARN  (main) [] o.e.j.x.XmlConfiguration Deprecated method public void org.eclipse.jetty.server.handler.gzip.GzipHandler.setDeflaterPoolCapacity(int) in file:///Users/risdenk/repos/apache/solr/solr/packaging/build/solr-10.0.0-SNAPSHOT/server/etc/jetty-gzip.xml
```

risdenk commented on solr/core/src/java/org/apache/solr/client/solrj/embedded/JettySolrRunner.java:
```diff
@@ -452,9 +457,6 @@ public void lifeCycleFailure(LifeCycle arg0, Throwable arg1) {
     gzipHandler.setHandler(chain);
 
     gzipHandler.setMinGzipSize(23); // https://github.com/eclipse/jetty.project/issues/4191
-    gzipHandler.setCheckGzExists(false);
-    gzipHandler.setCompressionLevel(-1);
-    gzipHandler.setExcludedAgentPatterns(".*MSIE.6\\.0.*");
```
* `setCheckGzExists​(boolean checkGzExists)` - removed w/o replacement - https://www.eclipse.org/jetty/javadoc/jetty-9/org/eclipse/jetty/server/handler/gzip/GzipHandler.html#setCheckGzExists(boolean)
* `setCompressionLevel` - removed from Jetty 10 - https://www.eclipse.org/jetty/javadoc/jetty-10/org/eclipse/jetty/server/handler/gzip/GzipHandler.html
* `setExcludedAgentPatterns` - already includes MSIE 6 by default - see description here: https://www.eclipse.org/jetty/javadoc/jetty-10/org/eclipse/jetty/server/handler/gzip/GzipHandler.html

risdenk replied:
setCompressionLevel addressed in 22cfe1da7d3c9b88c5210447fb664957e29d3ee1

risdenk commented on solr/core/src/java/org/apache/solr/client/solrj/embedded/JettySolrRunner.java:
```diff
@@ -304,7 +304,11 @@ private void init(int port) {
         configuration.setSecureScheme("https");
-        configuration.addCustomizer(new SecureRequestCustomizer());
+        SecureRequestCustomizer customizer = new SecureRequestCustomizer(false);
+        sslcontext.setSniRequired(false);
+        customizer.setSniHostCheck(false);
```
SNI is server name indication and as far as I've seen only ever required if you have multiple websites behind a single server - and the server needs to figure out which website for which TLS certificate. Basically the ability to host multiple sites with multiple TLS certificates.

JettySolrRunner either in tests or as an embedded server doesn't need to support SNI so makes SNI not required and doesn't do the SNI check.

elyograg replied:
It would be extremely unlikely that Solr would ever have more than one certificate chain installed, so I don't think we need to worry about SNI.  We actively discourage running more than one webapp in the same container as Solr, which I think would be about the only time you'd actually want SNI.  If somebody wants to get that fancy, they are on their own for getting Jetty to support what they need.

I'd really like to get to the point where we embed Jetty and manage the whole configuration ourselves.  Maybe have the Jetty configuration in solr.xml.  Then I think we could utilize bouncycastle to allow users to supply PEM files for TLS, to be converted on the fly at startup into what Java wants.  We are already including BC dependencies in Solr.

risdenk commented on solr/core/src/java/org/apache/solr/servlet/SolrRequestParsers.java:
```diff
@@ -670,9 +662,15 @@ public InputStream getStream() throws IOException {
   }
 
+  public static boolean isMultipart(HttpServletRequest req) {
+    // Jetty utilities
```
addressed in 7d799d816ddc70f17eff46368fa4e5048a9b3454

risdenk commented:
`setDeflaterPoolCapacity` addressed in 22cfe1da7d3c9b88c5210447fb664957e29d3ee1

dsmiley commented on solr/core/src/java/org/apache/solr/logging/LogWatcher.java:
```diff
@@ -150,7 +149,7 @@ private static LogWatcher<?> createWatcher(LogWatcherConfig config, SolrResource
 
     try {
-      slf4jImpl = StaticLoggerBinder.getSingleton().getLoggerFactoryClassStr();
+      slf4jImpl = LoggerFactory.getILoggerFactory().getClass().toString().replace("class ", "");
```
```suggestion
      slf4jImpl = LoggerFactory.getILoggerFactory().getClass().getName();
```

risdenk replied:
I was like this was dumb last night but couldn't figure out what I was missing. Thanks!

dsmiley commented on solr/core/src/java/org/apache/solr/util/StartupLoggingUtils.java:
```diff
@@ -50,7 +50,7 @@ public static void checkLogDir() {
 
   public static String getLoggerImplStr() { // nowarn
-    return binder.getLoggerFactoryClassStr();
+    return loggerFactory.getClass().toString().replace("class ", "");
```
```suggestion
    return loggerFactory.getClass().getName();
```

HoustonPutman reviewed (approved):
One question, one formatting thing.

Otherwise this looks good to go.

Btw we need to upgrade, as we are seeing https://github.com/eclipse/jetty.project/issues/8558 in our failing tests, and need an upgrade path when Jetty 10.0.13 is released.

risdenk commented on solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java:
```diff
@@ -278,6 +295,10 @@ public void close() {
   }
 
+  public void setAuthenticationStore(AuthenticationStore authenticationStore) {
+    this.authenticationStore.updateAuthenticationStore(authenticationStore);
```
good question - not sure. looks like we don't need this change. removed in 3ad2030756edb7e4a94e3745c8b19839c7ba5f0a

risdenk replied:
Well.... hit this error in tests after reverting this change:

```
  2> 61637 INFO  (TEST-Http2SolrClientTest.testSetCredentialsWithSysProps-seed#[6AB847A636F3DA61]) [] o.a.s.SolrTestCaseJ4 ###Ending testSetCredentialsWithSysProps
   >     java.lang.IllegalStateException
   >         at __randomizedtesting.SeedInfo.seed([6AB847A636F3DA61:9706EAF254309277]:0)
   >         at org.eclipse.jetty.client.HttpClient.setAuthenticationStore(HttpClient.java:327)
```

https://github.com/eclipse/jetty.project/blob/jetty-10.0.x/jetty-client/src/main/java/org/eclipse/jetty/client/HttpClient.java#L328

risdenk replied:
reverted again in a080a8c474314ff47ef922577691b52c2281b293 to allow authentication store to be set.

risdenk commented on solr/core/src/java/org/apache/solr/request/json/RequestUtil.java:
```diff
@@ -76,8 +76,13 @@ public static void processParams(
+
         String contentType = cs.getContentType();
-        if (contentType == null || !contentType.contains("/json")) {
+        if (contentType == null || (!contentType.contains("/json"))) {
```
more is better? I have no idea - removed now :) see 13cf43059e35ae26d298891e2682142342dce833

risdenk commented:
> Btw we need to upgrade, as we are seeing https://github.com/eclipse/jetty.project/issues/8558 in our failing tests, and need an upgrade path when Jetty 10.0.13 is released.

yup this also affects Jetty 9.4.x from what I gather from SOLR-16099

risdenk commented:
@markrmiller do you have any concerns on the current state of this PR? I'm hoping to merge it soon.

risdenk commented:
I haven't looked at these yet or see if they reproduce.

```
./gradlew check -Pvalidation.errorprone=true -Ptests.filter=@Nightly
...
ERROR: The following test(s) have failed:
  - org.apache.solr.hdfs.cloud.SharedFileSystemAutoReplicaFailoverTest.test (:solr:modules:hdfs)
    Test output: /Users/risdenk/repos/apache/solr/solr/modules/hdfs/build/test-results/test/outputs/OUTPUT-org.apache.solr.hdfs.cloud.SharedFileSystemAutoReplicaFailoverTest.txt
    Reproduce with: gradlew :solr:modules:hdfs:test --tests "org.apache.solr.hdfs.cloud.SharedFileSystemAutoReplicaFailoverTest.test" -Ptests.jvms=5 "-Ptests.jvmargs=-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=120m" -Ptests.seed=8633D060D495F1DF -Ptests.filter=@Nightly -Ptests.file.encoding=ISO-8859-1

  - org.apache.solr.cloud.api.collections.ShardSplitTest.testSplitAfterFailedSplit2 (:solr:core)
    Test output: /Users/risdenk/repos/apache/solr/solr/core/build/test-results/test/outputs/OUTPUT-org.apache.solr.cloud.api.collections.ShardSplitTest.txt
    Reproduce with: gradlew :solr:core:test --tests "org.apache.solr.cloud.api.collections.ShardSplitTest.testSplitAfterFailedSplit2" -Ptests.jvms=5 "-Ptests.jvmargs=-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=120m" -Ptests.seed=8633D060D495F1DF -Ptests.filter=@Nightly -Ptests.file.encoding=ISO-8859-1

  - org.apache.solr.cloud.api.collections.ShardSplitTest.testSplitWithChaosMonkey (:solr:core)
    Test output: /Users/risdenk/repos/apache/solr/solr/core/build/test-results/test/outputs/OUTPUT-org.apache.solr.cloud.api.collections.ShardSplitTest.txt
    Reproduce with: gradlew :solr:core:test --tests "org.apache.solr.cloud.api.collections.ShardSplitTest.testSplitWithChaosMonkey" -Ptests.jvms=5 "-Ptests.jvmargs=-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=120m" -Ptests.seed=8633D060D495F1DF -Ptests.filter=@Nightly -Ptests.file.encoding=ISO-8859-1

  - org.apache.solr.cloud.api.collections.ShardSplitTest.classMethod (:solr:core)
    Test output: /Users/risdenk/repos/apache/solr/solr/core/build/test-results/test/outputs/OUTPUT-org.apache.solr.cloud.api.collections.ShardSplitTest.txt
    Reproduce with: gradlew :solr:core:test --tests "org.apache.solr.cloud.api.collections.ShardSplitTest.classMethod" -Ptests.jvms=5 "-Ptests.jvmargs=-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=120m" -Ptests.seed=8633D060D495F1DF -Ptests.filter=@Nightly -Ptests.file.encoding=ISO-8859-1

  - org.apache.solr.cloud.api.collections.ShardSplitTest.classMethod (:solr:core)
    Test output: /Users/risdenk/repos/apache/solr/solr/core/build/test-results/test/outputs/OUTPUT-org.apache.solr.cloud.api.collections.ShardSplitTest.txt
    Reproduce with: gradlew :solr:core:test --tests "org.apache.solr.cloud.api.collections.ShardSplitTest.classMethod" -Ptests.jvms=5 "-Ptests.jvmargs=-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=120m" -Ptests.seed=8633D060D495F1DF -Ptests.filter=@Nightly -Ptests.file.encoding=ISO-8859-1
```

risdenk commented:
The above failures also happen on `main` branch so at least it is nothing new added from this PR for nightly tests.
